include/solvers/cahn_hilliard.h

```diff
@@ -0,0 +1,397 @@
+/* ---------------------------------------------------------------------
+ *
+ * Copyright (C) 2019 - 2020 by the Lethe authors
+ *
+ * This file is part of the Lethe library
+ *
+ * The Lethe library is free software; you can use it, redistribute
+ * it, and/or modify it under the terms of the GNU Lesser General
+ * Public License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ * The full text of the license can be found in the file LICENSE at
+ * the top level of the Lethe distribution.
+ *
+ * ---------------------------------------------------------------------
+
+ *
+ * Implementation of the Cahn-Hilliard equations as an auxiliary physics.
+ * Equations solved:
+ * dPhi/dt +  u * gradPhi =  div(M(Phi)*grad eta)
+ * eta - f(Phi) + epsilon^2 * div(grad Phi) = 0
+ * with Phi the phase field parameter (or phase order), eta the chemical
+ potential
+ * M the mobility function and epsilon the interface thickness
+ *
+ */
+
+#ifndef cahn_hilliard_h
+#define cahn_hilliard_h
+
+#include <core/bdf.h>
+#include <core/simulation_control.h>
+
+#include <solvers/auxiliary_physics.h>
+#include <solvers/cahn_hilliard_assemblers.h>
+#include <solvers/cahn_hilliard_scratch_data.h>
+#include <solvers/multiphysics_interface.h>
+
+#include <deal.II/base/convergence_table.h>
+#include <deal.II/base/quadrature_lib.h>
+
+#include <deal.II/distributed/solution_transfer.h>
+#include <deal.II/distributed/tria_base.h>
+
+#include <deal.II/fe/fe_q.h>
+#include <deal.II/fe/fe_simplex_p.h>
+#include <deal.II/fe/mapping_fe.h>
+#include <deal.II/fe/mapping_q.h>
+
+#include <deal.II/lac/trilinos_sparse_matrix.h>
+#include <deal.II/lac/trilinos_vector.h>
+
+
+template <int dim>
+class CahnHilliard : public AuxiliaryPhysics<dim, TrilinosWrappers::MPI::Vector>
+{
+public:
+  CahnHilliard<dim>(MultiphysicsInterface<dim> *     multiphysics_interface,
+                    const SimulationParameters<dim> &p_simulation_parameters,
+                    std::shared_ptr<parallel::DistributedTriangulationBase<dim>>
+                                                       p_triangulation,
+                    std::shared_ptr<SimulationControl> p_simulation_control)
+    : AuxiliaryPhysics<dim, TrilinosWrappers::MPI::Vector>(
+        p_simulation_parameters.non_linear_solver)
+    , multiphysics(multiphysics_interface)
+    , simulation_parameters(p_simulation_parameters)
+    , triangulation(p_triangulation)
+    , simulation_control(p_simulation_control)
+    , dof_handler(*triangulation)
+  {
+    if (simulation_parameters.mesh.simplex)
+      {
+        // for simplex meshes
+        const FE_SimplexP<dim> phase_order_fe(
+          simulation_parameters.fem_parameters.phase_ch_order);
+        const FE_SimplexP<dim> potential_fe(
+          simulation_parameters.fem_parameters.potential_order);
+        fe =
```

Comment:
Just a suggestion, but maybe it's not necessary: you could add a comment on how there's a difference between phase order and element order.